neighsyncd/neighsync.cpp

```diff
@@ -9,14 +9,276 @@
+    warmStartTimer(timespec{0, 0})
+{
+    WarmStart::checkWarmStart(m_appName, m_dockerName);
+
+    m_appTableName = m_appTable.getTableName();
+
+    // set the default timer value
+    if (defaultWarmStartTimerValue != 0)
+    {
+        m_reconcileTimer = defaultWarmStartTimerValue;
+    }
+
+    if (!WarmStart::isWarmStart())
+    {
+        m_warmStartInProgress = false;
+    }
+    else
+    {
+        m_warmStartInProgress = true;
+        uint32_t temp_value = WarmStart::getWarmStartTimer(m_appName, m_dockerName);
+        if (temp_value != 0)
+        {
+            m_reconcileTimer = temp_value;
+        }
+
+        warmStartTimer.setInterval(timespec{m_reconcileTimer, 0});
+
+        // Clear the producerstate table to make sure no pending data for the AppTable
+        m_psTable->clear();
+
+        WarmStart::setWarmStartState(m_appName, WarmStart::INIT);
+    }
+}
+
+AppRestartAssist::~AppRestartAssist()
+{
+}
+
+/* join the field-value strings for straight printing */
+string AppRestartAssist::joinVectorString(const vector<FieldValueTuple> &fv)
+{
+    string s;
+    for (const auto &temps : fv )
+    {
+	   s += temps.first + ":" + temps.second + ", ";
+    }
+    return s;
+}
+
+void AppRestartAssist::setCacheEntryState(std::vector<FieldValueTuple> &fvVector,
+    cache_state_t state)
+{
+    fvVector.back().second = cacheStateMap.at(state);
+}
+
+AppRestartAssist::cache_state_t AppRestartAssist::getCacheEntryState(const std::vector<FieldValueTuple> &fvVector)
+{
+    for (auto &iter : cacheStateMap)
+    {
+        if (fvVector.back().second == iter.second)
+        {
+            return iter.first;
+        }
+    }
+    throw std::logic_error("cache entry state is invalid");
+    return UNKNOWN;
+}
+
+/* Read table from APPDB and append stale flag then insert to cachemap */
+void AppRestartAssist::readTableToMap()
+{
+    vector<string> keys;
+
+    m_appTable.getKeys(keys);
+    FieldValueTuple state(CACHE_STATE_FIELD, "");
+
+    for (const auto &key: keys)
+    {
+        vector<FieldValueTuple> fv;
+
+	    /* if the fieldvalue is empty, skip */
+        if (!m_appTable.get(key, fv))
+        {
+            continue;
+        }
+
+        fv.push_back(state);
+        setCacheEntryState(fv, STALE);
+
+        string s = "";
+        s = joinVectorString(fv);
+
+        SWSS_LOG_INFO("write to cachemap: %s, key: %s, "
+               "%s", m_appTableName.c_str(), key.c_str(), s.c_str());
+
+        // insert to the cache map
+        appTableCacheMap[key] = fv;
+    }
+    WarmStart::setWarmStartState(m_appName, WarmStart::RESTORED);
+    SWSS_LOG_NOTICE("Restored appDB table to internal cache map");
+    return;
+}
+
+void AppRestartAssist::insertToMap(string key, vector<FieldValueTuple> fvVector, bool delete_key)
+{
+    string s;
+    s = joinVectorString(fvVector);
+
+    SWSS_LOG_INFO("Received message %s, key: %s, "
+            "%s, delete = %d", m_appTableName.c_str(), key.c_str(), s.c_str(), delete_key);
```

Comment:
s = joinVectorString(fv); is used for LOG_INFO only.
s._c_str() could be replaced with joinVectorString(fv).c_str(),